persist/fs/seek_manager.go

```diff
@@ -137,75 +170,198 @@ func (m *seekerManager) CacheShardIndices(shards []uint32) error {
 	return multiErr.FinalError()
 }
 
-func (m *seekerManager) openAnyUnopenSeekers(byTime *seekersByTime) error {
-	start := m.earliestSeekableBlockStart()
-	end := m.latestSeekableBlockStart()
-	blockSize := m.namespaceMetadata.Options().RetentionOptions().BlockSize()
-	multiErr := xerrors.NewMultiError()
+func (m *seekerManager) ConcurrentIDBloomFilter(shard uint32, start time.Time) (*ManagedConcurrentBloomFilter, error) {
+	byTime := m.seekersByTime(shard)
 
-	for t := start; !t.After(end); t = t.Add(blockSize) {
-		tNano := xtime.ToUnixNano(t)
-		byTime.RLock()
-		_, exists := byTime.seekers[tNano]
-		byTime.RUnlock()
+	// Try fast RLock() first
+	byTime.RLock()
+	startNano := xtime.ToUnixNano(start)
+	seekersAndBloom, ok := byTime.seekers[startNano]
+	byTime.RUnlock()
 
-		if exists {
-			// Avoid opening a new seeker if already an open seeker
-			continue
-		}
+	if ok && seekersAndBloom.wg == nil {
+		return seekersAndBloom.bloomFilter, nil
+	}
 
-		byTime.Lock()
-		_, exists = byTime.seekers[tNano]
-		if exists {
-			byTime.Unlock()
-			continue
-		}
+	byTime.Lock()
+	seekersAndBloom, err := m.getOrOpenSeekersWithLock(startNano, byTime)
+	byTime.Unlock()
+	return seekersAndBloom.bloomFilter, err
+}
 
-		seeker, err := m.newOpenSeeker(byTime.shard, t)
-		if err != nil {
-			if err != errSeekerManagerFileSetNotFound {
-				// Best effort to open files, if not there don't bother opening
-				multiErr = multiErr.Add(err)
-			}
-			byTime.Unlock()
-			continue
+func (m *seekerManager) Borrow(shard uint32, start time.Time) (ConcurrentFileSetSeeker, error) {
+	byTime := m.seekersByTime(shard)
+
+	byTime.Lock()
+	// Track accessed to precache in open/close loop
+	byTime.accessed = true
+
+	startNano := xtime.ToUnixNano(start)
+	seekersAndBloom, err := m.getOrOpenSeekersWithLock(startNano, byTime)
+	if err != nil {
+		return nil, err
```

Comment:
Hm, need to `byTime.Unlock()` here yeah?  Maybe we could pay the defer cost and `defer byTime.Unlock()` after locking here.

Reply:
yeah you're right good catch, I'll use defer